src/data/firebase.ts

```diff
@@ -0,0 +1,106 @@
+import { getImportedKey } from "./index";
+import * as firebase from "firebase/app";
+import "firebase/firestore";
+import { ExcalidrawElement } from "../element/types";
+import { getDrawingVersion } from "../element";
+
+const firebaseConfig = JSON.parse(process.env.REACT_APP_FIREBASE_CONFIG);
+
+firebase.initializeApp(firebaseConfig);
+
+interface FirebaseStoredDrawing {
+  drawingVersion: number;
+  ciphertext: firebase.firestore.Blob;
+}
+
+const EMPTY_IV = new Uint8Array(12);
+
+async function encryptElements(
+  key: string,
+  elements: readonly ExcalidrawElement[],
+): Promise<ArrayBuffer> {
+  const importedKey = await getImportedKey(key, "encrypt");
+  const iv = EMPTY_IV;
+  const json = JSON.stringify(elements);
+  const encoded = new TextEncoder().encode(json);
+  return await window.crypto.subtle.encrypt(
+    {
+      name: "AES-GCM",
+      iv,
+    },
+    importedKey,
+    encoded,
+  );
+}
+
+async function decryptElements(
+  key: string,
+  ciphertext: ArrayBuffer,
+): Promise<readonly ExcalidrawElement[]> {
+  const importedKey = await getImportedKey(key, "decrypt");
+  const decrypted = await window.crypto.subtle.decrypt(
+    {
+      name: "AES-GCM",
+      iv: EMPTY_IV,
+    },
+    importedKey,
+    ciphertext,
+  );
+
+  const decodedData = new TextDecoder("utf-8").decode(
+    new Uint8Array(decrypted) as any,
+  );
+  return JSON.parse(decodedData);
+}
+
+export async function saveToFirebase(
+  roomId: string,
+  roomSecret: string,
+  elements: readonly ExcalidrawElement[],
+) {
+  const drawingVersion = getDrawingVersion(elements);
+  const ciphertext = await encryptElements(roomSecret, elements);
+
+  const nextDocData = {
+    drawingVersion,
+    ciphertext: firebase.firestore.Blob.fromUint8Array(
+      new Uint8Array(ciphertext),
+    ),
+  } as FirebaseStoredDrawing;
+
+  const db = firebase.firestore();
+  const docRef = db.collection("drawings").doc(roomId);
+  const didUpdate = await db.runTransaction(async (transaction) => {
+    const doc = await transaction.get(docRef);
```

Comment:
I wonder if there's a way to do that without having to fetch the entire document. Does firebase have an API to update if newer by any chance?

Reply:
The documentation is not very useful. https://firebase.google.com/docs/firestore/transaction-data-contention#concurrency-controls

Reply:
We could just do `docRef.set(nextDocData)` and check the `drawingVersion` inside firestore security rule.

Reply:
someday maybe. for now let's stick with this. i want to keep the firebase rules super simple and i feel putting correctness semantics inside of firebase rules is scary (but maybe i'm just not used to how firebase rules are used)